smk/binning_preparation.smk

```diff
@@ -293,25 +303,45 @@ checkpoint make_assembly_batches:
     resources:
         mem = 5
     params:
-        min_length = lambda wildcards: wildcards.min_length
+        min_length = lambda wildcards: wildcards.min_length,
+        batch_filesize = calculate_batch_filesize
     run:
         import datetime
+        import os
 
         def logme(stream, msg):
             print(datetime.datetime.now().strftime("%d/%m/%Y %H:%M:%S"), msg, file=stream)
 
         with open(str(log), 'w') as f:
-            logme(f, "INFO: assemblies={} batch_size={}".format(len(input.assemblies), batch_size))
-            for i in range(0, len(input.assemblies), batch_size):
-                batch = 1 + int(i/batch_size)
-                logme(f, "INFO: Making BATCH={}".format(batch))
-
-                batch_input  = input.assemblies[i:i+batch_size]
-                batch_output = output.location + f"/batch{batch}.fasta.gz"
-                logme(f, "INFO:   INPUT ={}".format(batch_input))
-                logme(f, "INFO:   OUTPUT={}".format(batch_output))
-
-                filter_fasta_list_by_length(batch_input, batch_output, params.min_length)
+            logme(f, "INFO: assemblies={} batch_size={}Mb".format(len(input.assemblies), batch_filesize))
+            start = 0
+            batch = 1
+            current_batchfilesize = 0
+            for i in range(0, len(input.assemblies)):
+                current_assemblysize = os.path.getsize(input.assemblies[i]) / (1024**2)
+
+                # if adding the current assembly made batch larger than accepted, then write out batch
+                if (i and ((current_batchfilesize + current_assemblysize) > batch_filesize)) or (i+1 == len(input.assemblies)):
+                    logme(f, "INFO: Making BATCH={} size={:.0f}Mb".format(batch, current_batchfilesize))
+
+                    # take the final assembly too
+                    end = i
+                    if (i+1 == len(input.assemblies)):
```

Comment:
This could be a problem if the last one is a large assembly, and bwa index might take much more memory than the limit.  I changed the logic a little bit in my suggestion below.